torch/csrc/distributed/c10d/default_comm_hooks.cpp

```diff
@@ -0,0 +1,44 @@
+#include <torch/csrc/distributed/c10d/default_comm_hooks.h>
+
+#include <c10d/ProcessGroup.hpp>
+#include <torch/csrc/distributed/c10d/comm.h>
+#include <torch/torch.h>
+
+namespace c10d {
+
+c10::intrusive_ptr<torch::jit::Future> allReduceHook(
+    ProcessGroup* process_group,
+    GradBucket& bucket) {
+  auto allreduce_work = process_group->allreduce(*bucket.getTensorsRef());
+
+  auto div_by_process_group_size = [allreduce_work, process_group]() {
+    auto tensor = allreduce_work->result()[0] / process_group->getSize();
+    return c10::IValue(tensor);
+  };
+
+  auto fut = allreduce_work->getFuture();
+  return fut->then(div_by_process_group_size, fut->elementType());
+}
```

Comment:
We probably need to have tests for these hooks. Ideally these should be python tests once we hook up the entire structure to python land.

Reply:
I tested it locally by some hacking and reusing the same Python tests. The follow-up change is pybind the C++ implementation, and then we can add some tests.